### DISCUSSION TITLE
QueryNode Chunk Cache

### DISCUSSION THREAD

**Discussion by nairan-deshaw** [milvus-io/milvus#34296] [Q&A and General discussion] [answered]
We have a cluster deployment of Milvus on the k8s and our dataset sizes are of the order of a `150 million`. The querynodes are distributed across 16 replicas with `40G` of memory each. The current bottleneck for us in the partition load times and we exploring ways to improve the load times. We were referring to https://milvus.io/docs/chunk_cache.md and tried setting this up in our deployment. However, we do not see any noticeable improvements in the load times and from the logs, there is not indication that the chunk cache is being loaded. There are a couple of questions we have in this area:

1. Is chunk caching supposed to speed up partition load times in general or does it only come into play during vector search?
2. How to verify if chunk cache is working as expected?
3. Is the cache stored in the local storage of the querynodes under the `localStorage` directory which in our case happens to be `/var/lib/milvus/data` which is currently empty?
4. Which grafana panel can be used as an indication of checking the cache?
5. Which logs should specifically indicate that chunk cache is working?

The user.yaml file on the query nodes have the below configuration in them:
```sh
> kubectl exec -it qa-milvus-querynode -- cat /milvus/configs/user.yaml
Defaulted container "querynode" out of: querynode, config (init)
common:
  security:
    authorizationEnabled: true
proxy:
  maxUserNum: 500
  maxRoleNum: 100
queryNode:
  cache:
    enabled: true
    warmup: async
```

**Comment by yhmo** [comment]
As we know, we can specify "output_fields" of search() interface to output the original data. A normal load process only loads the vector index into memory, the original vector data is still in storage.
Without chunk cache, if you specify a vector field's name in the "output_fields", Milvus will have to read the original vector data from the S3.
With chunk cache, the load process will download the original vector data into the local disk, to achieve a better performance for the search request with output vectors.

You can try two cases:
1. search without "output_fields"
2. search with "output_fields" contains a vector field's name

**Comment by nairan-deshaw** [comment]
To clarify, when a `collection.load(partition_names=[partition_name])` is invoked, only the index of the collection is loaded in memory? How does it differ across partitions because we do see that the more the number of entries in the partition, the longer it takes to load the partition in memory.

I tried searching with both output and without output fields across multiple partitions of different sizes, here are the results:

1. With output_fields
```
Loaded partition_4 in 1.536184310913086s
Searched in 0.010303735733032227s
Released partition in 0.2179427146911621s
Loaded partition_5 in 3.046905517578125s
Searched in 0.017896413803100586s
Released partition in 0.6187968254089355s
Loaded partition_6 in 5.060715913772583s
Searched in 0.027882099151611328s
Released partition in 0.21604180335998535s
Loaded partition_7 in 4.558050870895386s
Searched in 0.029023408889770508s
Released partition in 0.42056846618652344s
```
2. Without output_fields:
```
Loaded partition_4 in 1.5356285572052002s
Searched in 0.006757259368896484s
Released partition in 0.4185295104980469s
Loaded partition_5 in 3.0476303100585938s
Searched in 0.012625455856323242s
Released partition in 0.41651391983032227s
Loaded partition_6 in 4.563213586807251s
Searched in 0.021347999572753906s
Released partition in 0.4178900718688965s
Loaded partition_7 in 4.560868263244629s
Searched in 0.019601106643676758s
Released partition in 0.4214146137237549s
```

Can you let me know where can I check to confirm if the chunk cache is being used since the loading times seem identical.

**Reply by yhmo** [reply]
- Milvus manages data by segments. For each segment, there is an independent index. Index progress is asynchronous.
- If a segment's index is not ready, the query node will load its original vector data into memory to do brute-force search
- Typically, the index size compare to the original vectors size: IVF_FLAT/HNSW, a bit larger than the original vectors size. IVF_SQ8/IVF_PQ, 25% ~ 30% of the original vectors size. DISKANN, 1/4 ~ 1/6 of the original vectors size.
- The time cost to load collection/partition mainly depends on how many segments, or how many indexes in this colleciton/partition
- Once the warmup is enabled, load process will download the original vector files to the local disk. Once the original vector files are already on local disk, the search with output_fileds will read data from the local files, instead of reading from S3.
- The time cost to output vector data depends how many vectors are read. If your search request NQ=1, topk=1, then there is only one vector is read from storage, doesn't affect search performance. Set topk=1000, then there will be NQ * 1000 vectors to be read, large impact to search performance.

**Reply by nairan-deshaw** [reply]
Thanks @yhmo for the details.

We are currently using a local minio instance within the k8s cluster.

To sum up my understanding here, when a search happens:
1. Load index from object storage to querynode memory
2. Load the original vector from object storage to memory if output_field is requested
3. If chunk cache is enabled, step 2 will just be returning the original vector from the local storage of the querynode

Is the above correct?

Couple of questions
- Can chunk cache include index files as  well?
- Which directory is the vector data stored on the querynodes?

As an extension, does enabling chunk cache then increase only the search performance since the index has to always be loaded into memory of the querynode whenever a load operation is invoked?

**Reply by yhmo** [reply]
Chunk cache is only for search with output_fields containing vector field, and not suitable for all index types.

Let's say there is a collection with two fields: "ID" and "Vector", the Vector field is already completely indexed. The collection has only one segment, the original vector data is stored in the S3 in this path:
**files/insert_log/[collection id]/[partition id]/[segment id]/[vector field id]/**
The index of the segment is stored in the S3 in this path:
**files/index_files/[index job id]/**

Now we call collection.load(). Milvus will load the index files into memory from S3 under the **files/index_files/[index job id]/**. If warmup is true, the original vector data files under the **files/insert_log/[collection id]/[partition id]/[segment id]/[vector field id]/** will be downloaded to the local disk, but these files are not loaded into memory.

The download path is configured in the milvus.yaml:
```
localStorage:
  path: /var/lib/milvus/data/
```
(Note that the files are invisible, we use mmap to map these files into memory)

Then we call **collection.search(data=data, anns_field="Vector", param=params, limit=topk, output_fields=["Vector"])**, note that we set output_fields=["Vector"] to return the original vector data of the similar items.

If index type is HNSW/IVF_FLAT, the index files already contain all the original vectors. So, milvus can read the original vectors from the in-memory index, no need to read the original vector data files from S3.
If index type is IVF_SQ8/IVF_PQ, the vectors are compressed in these indexes, milvus cannot know the original vectors from the index. So, milvus will have to read the original vector data files from S3.

If warmup is true, and the original vector data files are already downloaded to local, the search(output_fields=["Vector"]) performance will be much better since it doesn't need to read the files from remote S3.

So, we can know chunk cache doesn't help the performance of HNSW/IVF_FLAT. It helps improve the performance of IVF_SQ8/IVF_PQ when output_fields contain vector field.

Options for warmup: async and sync.
With "async", the query nodes download the original vector data files in the background, remote download process is very slow, may take a couple of minutes or even hours. Once after all the original data files are downloaded, the search performance of IVF_SQ8/IVF_PQ/DISKANN with output_fields=["Vector"] would be much better.
With "sync", the load process will wait untill all the original data files are downloaded, the load process will takes long times to return. Once the load process returns, the search performance of IVF_SQ8/IVF_PQ with output_fields=["Vector"] would be much better.


This is the entire story of chunk cache.
Conclusions:
1. chunk cache only helps IVF_SQ8/IVF_PQ/DISKANN performance with output_fields=["Vector"]
2. warmup=sync will increase the time cost of load process

**Comment by yhmo** [comment, marked as answer]
The time cost to load collection/partition mainly depends on the index file size.

Let's say you have 100GB data.
If you choose HNSW index, the total index size is 100GB+. When you call collection.load(), the query nodes will read 100GB+ from S3.
If you choose IVF_SQ8, the total index size is around 30GB. When you call collection.load(), the query nodes will read 30GB from S3, which is much faster.

Chunk cache doesn't help to reduce load time cost.

**Comment by nairan-deshaw** [comment]
Thanks for the detailed explanation @yhmo. Much appreciated!

**Reply by xiaofan-luan** [reply]
@tedxu  is working on optimizing performance of loading.

Generally speaking, loading only related to 3 things:

1. Your cpu usage and concurrency - tune parameter to increase load concurrency might help
2. Your S3/Minio throughput
3. Your data volume

**Comment by nairan-deshaw** [comment]
> tune parameter to increase load concurrency might help

Could you let me know which parameters can be tuned here?
